### ISSUE TITLE
Can we just replace PPO+RLHF with a preference models thats basically a transformer encoder + sigmoid model, trained with BCE. And during finetuning perform a reward maximization by just making the reward model predict 1s?

### ISSUE THREAD

**Issue by ssintelli** [lucidrains/PaLM-rlhf-pytorch#12]
At meta level, PPO based RLHF is performing minor adjustments to weights to align with human feedback. 

Can we just replace PPO+RLHF with a preference models thats basically a transformer encoder + sigmoid model, trained with BCE. And during finetuning perform a reward maximization by just making the reward model predict 1s? 

Sorry, if I am being naive. I do not have much experience with either RL or Large Language Models, but I would like to contribute to write a basic pytorch pipeline to do the following.

1. Train Transformer Encoder + Sigmoid Classifier on global token. (Preference Model Trainer)
2. Freeze Preference Model and Stick it on top of Large Language Model (assuming its end to end differentiable)
3. Maximize Preference, with gradient clipping on base model.

RLHFTrainer already implements large parts of 1 and 2. And Reward model is already in place. PaLM is already there. So, hope making changes to just maximize preference where we concatenate the input and response to reward model to predict all 1s wont be much difficult.

There are issues with end to end meaningful gradients backprop due to predictions being beam searched in infererence, but I hope we can fix that with some tricks.

**Comment by lucidrains** [comment]
do you mean you want the reward model to predict per token (say in the human feedback, one can give positive reward to a segment of response)? that part is in my plans

**Comment by ssintelli** [comment]
Replacing the core RL algorithm might be too far fetched. Instead of replacing the core RL based algorithm I will see if I can actually apply the above naive supervised end to end reward maximization as guide to PPO as auxiliary task in RLHF Trainer. Sorry if it sounds vague. I will try to work on it on weekends.  Also I need to go over all files in repo in details and recent literatures in incorporating human feedbacks to improve LLM so I might be completely wrong with my approach.

**Comment by lucidrains** [comment]
@ssintelli do you mean something like https://github.com/uber-research/PPLM ?

**Comment by lucidrains** [comment]
@ssintelli i feel like it is unlikely for any other approach to work, for the simple reason that if they did, OpenAI would not resort to something as complicated as PPO. they also have researchers like [Lilian Weng](https://lilianweng.github.io/posts/2021-01-02-controllable-text-generation/) who have considered all the possible options

**Comment by lucidrains** [comment]
@ssintelli haha, you read my deleted post

yea, let us know if you get PPLM working

**Comment by EricLee8** [comment]
I am also interested in your experiments, good luck and let me know if you get good results : )
